ddtrace/contrib/asgi/middleware.py

```diff
@@ -123,6 +123,7 @@ async def wrapped_send(message):
         except Exception:
             (exc_type, exc_val, exc_tb) = sys.exc_info()
             span.set_exc_info(exc_type, exc_val, exc_tb)
+            span.set_tag(http.STATUS_CODE, 500)
```

Comment:
Is the response code guaranteed to be 500 by the server if an error occurs? Just want to be sure that we're not going set a 500 status code when another response code could be returned.

Reply:
I have exactly this problem I had to intern the code and edit it to remove the span.finish from `finish` since I have a 

```python
try:
    await self.app(scope, receive, send_with_tracing)
except BaseException as exc:
    span.set_traceback()
    raise exc from None
span.finish()
```

And my Wrapper managing the status to raise

```python
from .middleware_ddtrace_asgi import TraceMiddleware
from ddtrace.ext import http as http_tags
from starlette.responses import Response
from starlette.types import Receive, Scope, Send
from fastapi.requests import Request

class ExceptionTraceMiddleware(TraceMiddleware):

    def __init__(self, **kwds):
        super().__init__(**kwds)

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return
        request = Request(scope=scope, receive=receive)
        try:
            await super(ExceptionTraceMiddleware, self).__call__(scope, receive, send)
        except errors.ApiError as exc:  # only our case
            response: Response = handle_api_error(request, exc)
            span = self.tracer.current_span()
            if span:
                span.set_tag(http_tags.STATUS_CODE, response.status_code)
                span.finish()
            await response.__call__(scope, receive, send)

```

Reply:
@raphaelauv By the looks of this code, I am guessing you are using the `ddtrace-asgi` library. But I think your solution is in line with what I have contributed here since the tag has to be set before the span is finished.

Reply:
Off course I'm using ddtrace-asgi take a look to the definition of my custom class , it's extending TraceMiddleware.

And No your solution is not equivalent because your putting 500 in all case and do not provide any mechanism to custom that value in case of error/exception.

Reply:
Thank you for highlighting this. I've added a static method on the middleware that can be overridden by a subclass to allow custom handling of any exceptions raised by the application.

https://github.com/DataDog/dd-trace-py/pull/1645/files#diff-61d5c837e13077f7c9d719d4cd911a59R79-R82

Hopefully that addresses your concerns.